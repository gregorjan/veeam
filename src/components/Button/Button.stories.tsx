import { Meta, StoryFn } from '@storybook/react'

import { Button } from './Button'

export default {
  component: Button,
  argTypes: {
    children: { control: 'text' },
    type: { control: 'radio', options: ['button', 'submit', 'reset'] },
  },
} as Meta<typeof Button>

export const Default: StoryFn<typeof Button> = (args) => <Button {...args} />
Default.storyName = 'Button'
Default.args = {
  children: 'button',
}
