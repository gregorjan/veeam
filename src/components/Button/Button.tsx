import { ComponentProps } from 'react'
import { styled } from '@linaria/react'
import { borderRadius, colors, rem, spacers } from '@styles'

interface StyledProps {
  variant?: 'primary' | 'secondary'
}
interface ButtonProps extends ComponentProps<'button'>, StyledProps {}

const StyledButton = styled.button<StyledProps>`
  appearance: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: none;
  cursor: pointer;
  text-transform: uppercase;
  text-decoration: none;
  white-space: nowrap;
  font-weight: 500;
  border-radius: ${borderRadius.sm};
  background-color: ${colors.transparent};
  padding: ${rem(spacers.spacer3)} ${rem(spacers.spacer4)};
  color: ${({ variant }) => (variant === 'primary' ? colors.white : colors.black)};
  background: ${({ variant }) => (variant === 'primary' ? colors.black : colors.transparent)};
  border: 1px solid ${({ variant }) => (variant === 'primary' ? colors.transparent : colors.black)};
`

export const Button = ({ children, type = 'button', variant = 'primary', ...props }: ButtonProps) => {
  return (
    <StyledButton type={type} variant={variant} {...props}>
      {children}
    </StyledButton>
  )
}
