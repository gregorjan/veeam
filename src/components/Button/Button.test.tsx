import { render } from '@testing-library/react'
import { Button } from './Button'

describe('Button', () => {
  it('renders', () => {
    const screen = render(<Button>button</Button>)
    const button = screen.getByText('button')
    expect(button).toBeInTheDocument()
  })
  describe('type prop', () => {
    it('defaults to type button', () => {
      const screen = render(<Button>button</Button>)
      const button = screen.getByText('button')
      expect(button).toHaveAttribute('type', 'button')
    })
    it('overrides to type button', () => {
      const screen = render(<Button type="button">button</Button>)
      const button = screen.getByText('button')
      expect(button).toHaveAttribute('type', 'button')
    })
    it('overrides to type submit', () => {
      const screen = render(<Button type="submit">button</Button>)
      const button = screen.getByText('button')
      expect(button).toHaveAttribute('type', 'submit')
    })
    it('overrides to type reset', () => {
      const screen = render(<Button type="reset">button</Button>)
      const button = screen.getByText('button')
      expect(button).toHaveAttribute('type', 'reset')
    })
  })
  describe('variant prop', () => {
    it('defaults to primary variant', () => {
      const screen = render(<Button>button</Button>)
      const button = screen.getByText('button')
      expect(button).toMatchSnapshot()
    })
    it('overides to primary variant', () => {
      const screen = render(<Button variant="primary">button</Button>)
      const button = screen.getByText('button')
      expect(button).toMatchSnapshot()
    })
    it('overides to secondary variant', () => {
      const screen = render(<Button variant="secondary">button</Button>)
      const button = screen.getByText('button')
      expect(button).toMatchSnapshot()
    })
  })
})
