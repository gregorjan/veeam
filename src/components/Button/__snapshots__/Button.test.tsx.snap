// Vitest Snapshot v1

exports[`Button > variant prop > defaults to primary variant 1`] = `
<button
  class="ssc0o8j"
  style="--ssc0o8j-0: white; --ssc0o8j-1: #111; --ssc0o8j-2: rgba(0, 0, 0, 0);"
  type="button"
>
  button
</button>
`;

exports[`Button > variant prop > overides to primary variant 1`] = `
<button
  class="ssc0o8j"
  style="--ssc0o8j-0: white; --ssc0o8j-1: #111; --ssc0o8j-2: rgba(0, 0, 0, 0);"
  type="button"
>
  button
</button>
`;

exports[`Button > variant prop > overides to secondary variant 1`] = `
<button
  class="ssc0o8j"
  style="--ssc0o8j-0: #111; --ssc0o8j-1: rgba(0, 0, 0, 0); --ssc0o8j-2: #111;"
  type="button"
>
  button
</button>
`;
